torch/distributed/fsdp/_symbolic_trace.py

```diff
@@ -0,0 +1,241 @@
+                execution_info._traced_param_set.add(p)
+    return proxy
+
+
+def _patched_call_module(
+    call_module: Callable,
+    execution_info: _ExecutionInfo,
+    module: torch.nn.Module,
+    forward: Callable[..., Any],
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any],
+) -> Any:
+    """
+    Override of :meth:`~torch.fx.Tracer.call_module`.
+    ``Tracer.call_module`` is called in symbolic tracing for each non-root module.
+    This override intercepts the recording of each operation to
+    update ``execution_info.module_forward_order`` and ``execution_info.module_to_execution_infos``.
+
+    Args:
+        call_module (Callable):
+            The ``call_module`` function to be patched.
+        execution_info (_ExecutionInfo):
+            Used to repord the execution information.
+        module (torch.nn.Module):
+            The module for which a call is being emitted.
+        forward (Callable[..., Any]):
+            The ``forward()`` method of the ``torch.nn.Module`` to be invoked.
+        args (Tuple[Any, ...]):
+            ``args`` of the module callsite.
+        kwargs (Dict[str, Any]):
+            ``kwargs`` of the module callsite.
+    """
+    execution_info.module_forward_order.append(module)
+    named_params = list(module.named_parameters())
+    if named_params:
+        execution_info.module_to_execution_infos[execution_info.current_module].append(
+            (module, list(module.named_parameters()))
+        )
+    # Stores away current_module for restoration later
+    prev_current_module = execution_info.current_module
+    execution_info.current_module = module
+    # Note that if the forward of module is called multiple times, this will record
+    # the execution info of the last forward pass.
+    execution_info.module_to_execution_infos[module] = []
+    output = call_module(module, forward, args, kwargs)
+    execution_info.current_module = prev_current_module
+    return output
+
+
+@contextlib.contextmanager
+def _patch_tracer(
+    tracer: torch.fx.Tracer,
+    root_module: torch.nn.Module,
+    execution_info: _ExecutionInfo,
+) -> Generator:
+    """
+    Within the context manager, patches the input tracer so that during
+    ``tracer.trace()``, the forward order of all modules and the parameter
+    execution information are recorded. The patches of the input tracer
+    will be removed after the context manager exits.
+
+    Args:
+        tracer (torch.fx.Tracer): the input ``tracer`` whose member functions
+            will be patched within the context manager.
+        root_module (torch.nn.Module): the top-level module to be traced
+            and should not contain any FSDP modules.
+        execution_info (_ExecutionInfo): used to record the execution order information
+            when performing ``tracer.trace()`` within the context manager.
+
+    """
+    from .fully_sharded_data_parallel import FullyShardedDataParallel
```

Comment:
Any Python experts know how to best resolve this kind of import problem? 😆 
I am assuming we are importing here to avoid circular imports. I wonder if there is any runtime consequence for doing this inside a method that may be called more than once.

Reply:
yes this is for avoiding circular imports. I'm not sure what's the best way for that

Reply:
changed based on Rohan's suggestion